x-pack/platform/plugins/shared/onechat/server/services/tools/builtin/definitions/create_visualization/create_visualization.ts

```diff
@@ -0,0 +1,146 @@
+  existingConfig: z
+    .string()
+    .optional()
+    .describe('An existing visualization configuration to modify.'),
+  chartType: z
+    .enum([SupportedChartType.Metric, SupportedChartType.Map])
+    .optional()
+    .describe(
+      '(optional) The type of chart to create as indicated by the user. If not provided, the LLM will suggest the best chart type.'
+    ),
+  esql: z
+    .string()
+    .optional()
+    .describe(
+      '(optional) An ES|QL query. If not provided, tool with automatically generate the query.'
+    ),
+});
+
+export const createVisualizationTool = (): BuiltinToolDefinition<
+  typeof createVisualizationSchema
+> => {
+  return {
+    id: platformCoreTools.createVisualization,
+    type: ToolType.builtin,
+    description: `Create a visualization configuration based on a natural language description.
+
+This tool will:
+1. Determine the best chart type if not specified (from: ${Object.values(SupportedChartType).join(
+      ', '
+    )})
+2. Generate an ES|QL query if not provided
+3. Generate a valid visualization configuration`,
+    schema: createVisualizationSchema,
+    tags: [],
+    handler: async (
+      { query: nlQuery, chartType, esql, existingConfig },
+      { esClient, modelProvider, logger, events }
+    ) => {
+      try {
+        // Step 1: Determine chart type if not provided
+        let selectedChartType: SupportedChartType = chartType || SupportedChartType.Metric;
+        const parsedExistingConfig = existingConfig ? JSON.parse(existingConfig) : null;
+
+        if (!chartType) {
+          logger.debug('Chart type not provided, using LLM to suggest one');
+          const suggestedChartType = await getChartType(
+            modelProvider,
+            parsedExistingConfig?.type,
+            nlQuery
+          );
+          if (
+            Object.values(SupportedChartType).includes(suggestedChartType as SupportedChartType)
+          ) {
+            selectedChartType = suggestedChartType as SupportedChartType;
+          } else {
+            logger.warn(
+              `LLM suggested invalid chart type: ${suggestedChartType}, defaulting to metric`
+            );
+            selectedChartType = SupportedChartType.Metric;
+          }
+        }
+
+        // Step 2: Generate visualization configuration using langgraph with validation retry
+        const model = await modelProvider.getDefaultModel();
+        const schema = metricSchema;
+
+        // Create and invoke the validation retry graph
+        const graph = createVisualizationGraph(model, logger, events, esClient);
+
+        const finalState = await graph.invoke({
+          nlQuery,
+          chartType: selectedChartType,
+          schema,
+          existingConfig,
+          parsedExistingConfig,
+          esqlQuery: esql || '',
+          currentAttempt: 0,
+          actions: [],
+          validatedConfig: null,
+          error: null,
+        });
```

Comment:
I think we need to configure the `recursionLimit` according to the value of `MAX_RETRY_ATTEMPTS`. `recursionLimit` defaults to 10, meaning that we would hit the limit before the 5th retry atm (maybe even before the 4th)

Can just hardcode a higher value for now, e.g.

https://github.com/elastic/kibana/blob/ac1a0f576e29fd34b50b7f67176494694e48c91f/x-pack/platform/plugins/shared/onechat/server/services/agents/modes/default/run_chat_agent.ts#L99